Services/communication_service.c

```diff
@@ -35,31 +53,753 @@
+        U_stat.firmware_ver = csp_hton32((uint32_t) U_stat.firmware_ver);
+        U_stat.payload_size = csp_hton16(U_stat.payload_size);
+        U_stat.secure_key = csp_hton32(U_stat.secure_key);
+
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        memcpy(&packet->data[OUT_DATA_BYTE], &U_stat, sizeof(U_stat));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(U_stat) + 1);
+        if (queue_response(packet) != SATR_OK)
+        {
+            return SATR_ERROR;
+        }
+        break;
+
+    case UHF_GET_CALL_SIGN:
+        status =  HAL_UHF_getDestination(&U_callsign.dest)
+                + HAL_UHF_getSource(&U_callsign.src);
+
+        uhf_struct_len = 6;
+        //uhf_struct_len = U_callsign.dest.len; //if different, use this and separate loops
+        //uhf_struct_len = U_callsign.src.len;
+        uint8_t dst[24]; //size of unicode char = 4
+        uint8_t src[24];
+        memset(dst,0,24); // Using {0} not recommended, gives warnings and possible errors.
+        memset(src,0,24);
+        if (sizeof(dst) + sizeof(src) + 1 > csp_buffer_data_size())
+                {
+                    return SATR_ERROR;
+                }
+        for (i = 0; i < uhf_struct_len; i++)
+                {
+                    dst[3+4*i] = csp_hton32((uint32_t) U_callsign.dest.message[i]);
+                    src[3+4*i] = csp_hton32((uint32_t) U_callsign.src.message[i]);
+                }
+
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        memcpy(&packet->data[OUT_DATA_BYTE], &dst, sizeof(dst));
+        memcpy(&packet->data[OUT_DATA_BYTE + sizeof(dst)], &src, sizeof(src));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(dst) + sizeof(src)  + 1);
+        if (queue_response(packet) != SATR_OK)
+        {
+            return SATR_ERROR;
+        }
+        break;
+
+    case UHF_GET_MORSE:
+        status = HAL_UHF_getMorse(&U_beacon.morse);
+
+        uhf_struct_len = 36;
+        //uhf_struct_len = U_beacon.morse.len;
+        uint8_t mrs[144]; //size of unicode char = 4
+        memset(mrs,0,144);
+        if (sizeof(mrs) + 1 > csp_buffer_data_size())
+                {
+                    return SATR_ERROR;
+                }
+        for (i = 0; i < uhf_struct_len; i++)
+                {
+                    mrs[3+4*i] = csp_hton32(
+                            (uint32_t) U_beacon.morse.message[i]);
+                }
+
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        memcpy(&packet->data[OUT_DATA_BYTE], &mrs, sizeof(mrs));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(mrs) + 1);
+        if (queue_response(packet) != SATR_OK)
+        {
+            return SATR_ERROR;
+        }
+        break;
+
+    case UHF_GET_MIDI:
+        status = HAL_UHF_getMIDI(&U_beacon.MIDI);
+
+        uhf_struct_len = 36;
+        //uhf_struct_len = U_beacon.midi.len;
```

Comment:
So, why not store the length along with the value in the struct for each of these things? Could we set the .len property to the required value when the value is set and then use the .len property when retreiving the value, and then if we were to decide one day to put a length feild in the ground command then we would only have to change the setters, if you see what I mean?

Reply:
We do store the length in the struct in the set functions. With malloc, we should be able to use the commented-out line. Having two options for the ground command would make things ugly I guess. Let me see if I can come up with a proper way of doing that.